File: sample/go/08_langtons_ant.go
package main


// 08: Sample that outputs Langton's Ant trajectories as a GIF.

func capture(grid []any, w int64, h int64) []any {
    var frame []any = __pytra_as_list(__pytra_bytearray((w * h)))
    for y := int64(0); y < h; y += 1 {
        var row_base int64 = (y * w)
        for x := int64(0); x < w; x += 1 {
            __pytra_set_index(frame, (row_base + x), __pytra_ifexp((__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(grid, y)), x)) != 0), int64(255), int64(0)))
        }
    }
    return __pytra_as_list(__pytra_bytes(frame))
}

func run_08_langtons_ant() {
    var w int64 = int64(420)
    var h int64 = int64(420)
    var out_path string = __pytra_str("sample/out/08_langtons_ant.gif")
    var start float64 = __pytra_perf_counter()
    var grid []any = __pytra_as_list(func() []any { __out := []any{}; __step := __pytra_int(int64(1)); for __lc_i := __pytra_int(int64(0)); (__step >= 0 && __lc_i < __pytra_int(h)) || (__step < 0 && __lc_i > __pytra_int(h)); __lc_i += __step { __out = append(__out, __pytra_list_repeat(int64(0), w)) }; return __out }())
    var x int64 = __pytra_int((w / int64(2)))
    var y int64 = __pytra_int((h / int64(2)))
    var d int64 = int64(0)
    var steps_total int64 = int64(600000)
    var capture_every int64 = int64(3000)
    var frames []any = __pytra_as_list([]any{})
    for i := int64(0); i < steps_total; i += 1 {
        if (__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(grid, y)), x)) == int64(0)) {
            d = ((d + int64(1)) % int64(4))
            __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, y)), x, int64(1))
        } else {
            d = ((d + int64(3)) % int64(4))
            __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, y)), x, int64(0))
        }
        if (d == int64(0)) {
            y = (((y - int64(1)) + h) % h)
        } else {
            if (d == int64(1)) {
                x = ((x + int64(1)) % w)
            } else {
                if (d == int64(2)) {
                    y = ((y + int64(1)) % h)
                } else {
                    x = (((x - int64(1)) + w) % w)
                }
            }
        }
        if ((i % capture_every) == int64(0)) {
            frames = append(frames, capture(grid, w, h))
        }
    }
    __pytra_save_gif(out_path, w, h, frames, __pytra_grayscale_palette(), int64(5), int64(0))
    var elapsed float64 = (__pytra_perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", __pytra_len(frames))
    __pytra_print("elapsed_sec:", elapsed)
}

func main() {
    run_08_langtons_ant()
}

File: sample/swift/08_langtons_ant.swift
import Foundation


// 08: Sample that outputs Langton's Ant trajectories as a GIF.

func capture(grid: [Any], w: Int64, h: Int64) -> [Any] {
    var frame: [Any] = __pytra_as_list(__pytra_bytearray((__pytra_int(w) * __pytra_int(h))))
    let __step_0 = __pytra_int(Int64(1))
    var y = __pytra_int(Int64(0))
    while ((__step_0 >= 0 && y < __pytra_int(h)) || (__step_0 < 0 && y > __pytra_int(h))) {
        var row_base: Int64 = __pytra_int((__pytra_int(y) * __pytra_int(w)))
        let __step_1 = __pytra_int(Int64(1))
        var x = __pytra_int(Int64(0))
        while ((__step_1 >= 0 && x < __pytra_int(w)) || (__step_1 < 0 && x > __pytra_int(w))) {
            __pytra_setIndex(frame, (__pytra_int(row_base) + __pytra_int(x)), __pytra_ifexp((__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x)) != 0), Int64(255), Int64(0)))
            x += __step_1
        }
        y += __step_0
    }
    return __pytra_bytes(frame)
}

func run_08_langtons_ant() {
    var w: Int64 = __pytra_int(Int64(420))
    var h: Int64 = __pytra_int(Int64(420))
    var out_path: String = __pytra_str("sample/out/08_langtons_ant.gif")
    var start: Double = __pytra_float(__pytra_perf_counter())
    var grid: [Any] = __pytra_as_list(({ () -> [Any] in var __out: [Any] = []; let __step = __pytra_int(Int64(1)); var __lc_i = __pytra_int(Int64(0)); while ((__step >= 0 && __lc_i < __pytra_int(h)) || (__step < 0 && __lc_i > __pytra_int(h))) { __out.append(__pytra_list_repeat(Int64(0), w)); __lc_i += __step }; return __out })())
    var x: Int64 = __pytra_int((__pytra_int(__pytra_int(w) / __pytra_int(Int64(2)))))
    var y: Int64 = __pytra_int((__pytra_int(__pytra_int(h) / __pytra_int(Int64(2)))))
    var d: Int64 = __pytra_int(Int64(0))
    var steps_total: Int64 = __pytra_int(Int64(600000))
    var capture_every: Int64 = __pytra_int(Int64(3000))
    var frames: [Any] = __pytra_as_list([])
    let __step_0 = __pytra_int(Int64(1))
    var i = __pytra_int(Int64(0))
    while ((__step_0 >= 0 && i < __pytra_int(steps_total)) || (__step_0 < 0 && i > __pytra_int(steps_total))) {
        if (__pytra_int(__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x))) == __pytra_int(Int64(0))) {
            d = __pytra_int((__pytra_int((__pytra_int(d) + __pytra_int(Int64(1)))) % __pytra_int(Int64(4))))
            __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x, Int64(1))
        } else {
            d = __pytra_int((__pytra_int((__pytra_int(d) + __pytra_int(Int64(3)))) % __pytra_int(Int64(4))))
            __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x, Int64(0))
        }
        if (__pytra_int(d) == __pytra_int(Int64(0))) {
            y = __pytra_int((__pytra_int((__pytra_int((__pytra_int(y) - __pytra_int(Int64(1)))) + __pytra_int(h))) % __pytra_int(h)))
        } else {
            if (__pytra_int(d) == __pytra_int(Int64(1))) {
                x = __pytra_int((__pytra_int((__pytra_int(x) + __pytra_int(Int64(1)))) % __pytra_int(w)))
            } else {
                if (__pytra_int(d) == __pytra_int(Int64(2))) {
                    y = __pytra_int((__pytra_int((__pytra_int(y) + __pytra_int(Int64(1)))) % __pytra_int(h)))
                } else {
                    x = __pytra_int((__pytra_int((__pytra_int((__pytra_int(x) - __pytra_int(Int64(1)))) + __pytra_int(w))) % __pytra_int(w)))
                }
            }
        }
        if (__pytra_int((__pytra_int(i) % __pytra_int(capture_every))) == __pytra_int(Int64(0))) {
            frames = __pytra_as_list(frames); frames.append(capture(grid, w, h))
        }
        i += __step_0
    }
    __pytra_noop(out_path, w, h, frames, [])
    var elapsed: Double = __pytra_float((__pytra_float(__pytra_perf_counter()) - __pytra_float(start)))
    __pytra_print("output:", out_path)
    __pytra_print("frames:", __pytra_len(frames))
    __pytra_print("elapsed_sec:", elapsed)
}

@main
struct Main {
    static func main() {
        run_08_langtons_ant()
    }
}
